packages/http/test/file.test.ts

```diff
@@ -0,0 +1,547 @@
+import { expectDiagnosticEmpty, expectDiagnostics } from "@typespec/compiler/testing";
+import { deepStrictEqual, ok, strictEqual } from "node:assert";
+import { describe, it } from "vitest";
+import { isHeader } from "../src/decorators.js";
+import { compileOperationsFull, getOperations } from "./test-host.js";
+
+/*
+
+CASES TO TEST
+
+// `"single"` body: uploads file as a JSON structure { filename: <string>,
+// contentType: <string>, contents: <base64> }
+op uploadFileAsJson(@header contentType: "application/json", @body file: File)
+
+// `"file"` body: Uploads file as a binary data stream. Content-Type has contentType and
+// filename cannot be specified (Content-Disposition is a response-only header)
+op uploadFile(@body file: File);
+op uploadFile(...File);
+
+// `"file"` body: Downloads file as binary data stream. Content-Type has contentType and
+// filename comes from Content-Disposition
+op downloadFile(): { @body file: File };
+op downloadFile(): File;
+
+
+// Declares a custom textual file that can either be "application/json" or "application/yaml"
+model SpecFile extends File<Contents = bytes> {
+  contentType: "application/json" | "application/yaml";
```

Comment:
if the `contentType` is `application/json` only, and there is no operation level content type header, it still means that i need to upload a json file, right?

Reply:
Like

```tsp
model JsonFile extends File {
  contentType: "application/json";
}

@post op upload(@body file: JsonFile): void;
```

?

This means you need to upload a file, and the _contents_ of the file must be JSON. This will be a request like

```
POST /
Content-Type: application/json
Content-Length: <file size>

<file contents, which must be JSON>
```

If you instead say

```tsp
@post op upload(@header contentType: "application/json", @body file: JsonFile): void;
```

you will get a request like

```
POST /
Content-Type: "application/json"

{"contentType":<string?>,"filename":<string?>,"contents":<base64 file contents, which must also be valid JSON>}
```